File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void backtrack(int open,int close,int &n,vector<string> &ans,string current){
        if(open==n && close==n ){
            ans.push_back(current);
            return;
            }
            if(open<n){
                backtrack(open+1,close,n,ans,current+'(');
            }
            if(close <open){
                backtrack(open,close+1,n,ans,current+')');
            }
    }
    vector<string> generateParenthesis(int n) {
        vector<string>ans;
        backtrack(0,0,n,ans,"");
        return ans;
        
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution(object):
    def backtrack(self,n,open,close,ans,possiblestring):
        if open==n and close ==n:
            ans.append(possiblestring)
            return
        if(open<n):
            self.backtrack(n,open+1,close,ans,possiblestring+'(')
        if(close<open) :
            self.backtrack(n,open,close+1,ans,possiblestring+')')   


    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        ans=[]
        self.backtrack(n,0,0,ans,"")
        return ans
        